app/routes.py

```diff
@@ -1,2 +1,239 @@
+             
+            response_body = {
+                        "id": goal.goal_id,
+                        "task_ids": task_ids
+                    }
+                    
+            return jsonify(response_body), 200
+
+        if request.method == "GET":
+            tasks_response =[]
+            for task in goal.tasks:
+                tasks_response.append({
+                    "id": task.task_id,
+                    "goal_id": task.goal_id,
+                    "title": task.title,
+                    "description": task.description,
+                    "is_complete": bool(task.completed_at)
+                })
+            response_body = {
+                "id": goal.goal_id,
+                "title": goal.title,
+                "tasks" : tasks_response
+            }
+            return jsonify(response_body), 200
+
+
+@tasks_bp.route("",methods=["GET","POST"])
+def handle_task():
+    if request.method == "GET": 
+
+        sorting_task = request.args.get('sort') 
+        list = None
+        if sorting_task == "desc":
+            list = Task.query.order_by(Task.title.desc()) # descending method
+        elif sorting_task == "asc":
+            list = Task.query.order_by(Task.title.asc()) # ascending method
+        else: 
+            list = Task.query.all()
+        tasks_response = []
+        for task in list:
+            tasks_response.append({
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete" : False
+            if task.completed_at == None else True
+            })
+
+        return jsonify(tasks_response)
+
+    elif request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body or "description" not in request_body or "completed_at" not in request_body:
+            return{
+                "details": "Invalid data"
+            },400
+
+        new_task = Task (
+            title=request_body["title"],
+            description=request_body["description"],
+            completed_at=request_body["completed_at"]
+        )
+
+        db.session.add(new_task)
+        db.session.commit()
+
+        return {
+            "task":new_task.task_dict()
+        }, 201
+
+def update_completion(task_id, value):
+    task = Task.query.get(task_id)
+    
+    if not task:
+        return("Task not found",404)
+    
+    task.completed_at = value 
+
+    return {
+        "task":task.task_dict()
+    }, 200 
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    return update_completion(task_id, datetime.now())
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods =["PATCH"])
+def mark_incomplete(task_id):
+    return update_completion(task_id,None)
+
+
+@tasks_bp.route("/<task_id>", methods=["GET","PUT","DELETE"])
+def handle_task_get(task_id):
+    task = Task.query.get(task_id)
+    if task == None:
+        return ("", 404)
```

Comment:
Notice that this chunk of looking up an id, checking for None, and returning a 404 when not found recurs frequently. It's a little tricky to extract this into a helper function, mostly since a helper function can't force the calling function to exit. That is, unless it raises an error. So we could think about writing a helper function to perform this set of steps, but we would also need to customize the Flask error handling behavior to return JSON rather than HTML. There _is_ the `get_or_404` method that essentially does the lookup with errors already, but we would should still update the flask error handling (see [here](https://flask.palletsprojects.com/en/2.0.x/errorhandling/#error-handlers) for more details.)

Another approach could be to use a decorator to intercept the task id from the route, and performt the lookup before reaching the body of our endpoint. There's an example of that [here](https://github.com/AdaGold/hello-books-api-decorated/blob/main/app/routes.py).